core/src/main/java/org/hisp/dhis/android/core/trackedentity/TrackedEntityAttributeReservedValueManager.kt

```diff
@@ -0,0 +1,409 @@
+import org.hisp.dhis.android.core.organisationunit.OrganisationUnitProgramLinkTableInfo
+import org.hisp.dhis.android.core.organisationunit.internal.OrganisationUnitProgramLinkStore
+import org.hisp.dhis.android.core.organisationunit.internal.OrganisationUnitStore
+import org.hisp.dhis.android.core.program.ProgramTrackedEntityAttributeTableInfo
+import org.hisp.dhis.android.core.program.internal.ProgramTrackedEntityAttributeStore
+import org.hisp.dhis.android.core.settings.GeneralSettingObjectRepository
+import org.hisp.dhis.android.core.trackedentity.internal.ReservedValueSettingStore
+import org.hisp.dhis.android.core.trackedentity.internal.TrackedEntityAttributeReservedValueQuery
+import org.hisp.dhis.android.core.trackedentity.internal.TrackedEntityAttributeReservedValueStore
+import org.hisp.dhis.android.core.trackedentity.internal.TrackedEntityAttributeStore
+import org.hisp.dhis.android.core.user.internal.UserOrganisationUnitLinkStore
+
+@SuppressWarnings("LongParameterList", "TooManyFunctions")
+@Reusable
+class TrackedEntityAttributeReservedValueManager @Inject internal constructor(
+    private val store: TrackedEntityAttributeReservedValueStore,
+    private val organisationUnitStore: OrganisationUnitStore,
+    private val trackedEntityAttributeStore: TrackedEntityAttributeStore,
+    private val programTrackedEntityAttributeStore: ProgramTrackedEntityAttributeStore,
+    private val organisationUnitProgramLinkStore: OrganisationUnitProgramLinkStore,
+    private val userOrganisationUnitLinkStore: UserOrganisationUnitLinkStore,
+    private val generalSettingObjectRepository: GeneralSettingObjectRepository,
+    private val reservedValueSettingStore: ReservedValueSettingStore,
+    private val executor: D2CallExecutor,
+    private val reservedValueQueryCallFactory: QueryCallFactory<TrackedEntityAttributeReservedValue,
+        TrackedEntityAttributeReservedValueQuery>
+) {
+    private val d2ProgressManager = D2ProgressManager(null)
+
+    /**
+     * @param attributeUid        Attribute uid
+     * @param organisationUnitUid Organisation unit uid
+     * @return Value of tracked entity attribute
+     * @see .getValue
+     */
+    fun blockingGetValue(attributeUid: String, organisationUnitUid: String): String {
+        return runBlocking {
+            getValueCoroutines(attributeUid, organisationUnitUid) ?: ""
+        }
+    }
+
+    /**
+     * Get a reserved value and remove it from database. If the number of available values is below a threshold
+     * (default [.FILL_UP_TO] * [.FACTOR_TO_REFILL]) it tries to download before returning a value.
+     *
+     * @param attributeUid        Attribute uid
+     * @param organisationUnitUid Organisation unit uid
+     * @return Single with value of tracked entity attribute
+     */
+
+    fun getValue(attributeUid: String, organisationUnitUid: String): Single<String> {
+        return rxSingle { getValueCoroutines(attributeUid, organisationUnitUid) ?: "" }
```

Comment:
Previously, this method could return:
- an exception (if no more reserved values)
- the reserved value
- a null value (if for some reason there is a reservedValue with null value)

Now, the third case is defaulted to `""`, which changes the behavior and a sense that the value is correct, which is not right. It would be better to throw an exception or to return a null value. If we want to keep it the return value to strict NonNull, we can throw an exception instead

Reply:
If `getValueCoroutines` returns NonNull, the expression `?: ""` is not needed anymore, is it?